File: api/src/main/java/com/openlattice/edm/collection/CollectionTemplateType.java
package com.openlattice.edm.collection;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.openlattice.client.serialization.SerializationConstants;

import java.util.Objects;
import java.util.Optional;
import java.util.UUID;

public class CollectionTemplateType {

    private final UUID   id;
    private       String name;
    private       String title;
    private       String description;
    private final UUID   entityTypeId;

    @JsonCreator
    public CollectionTemplateType(
            @JsonProperty( SerializationConstants.ID_FIELD ) Optional<UUID> id,
            @JsonProperty( SerializationConstants.NAME ) String name,
            @JsonProperty( SerializationConstants.TITLE_FIELD ) String title,
            @JsonProperty( SerializationConstants.DESCRIPTION_FIELD ) Optional<String> description,
            @JsonProperty( SerializationConstants.ENTITY_TYPE_ID ) UUID entityTypeId ) {
        this.id = id.orElse( UUID.randomUUID() );
        this.name = name;
        this.title = title;
        this.description = description.orElse( "" );
        this.entityTypeId = entityTypeId;
    }

    public CollectionTemplateType(
            UUID id,
            String name,
            String title,
            Optional<String> description,
            UUID entityTypeId ) {
        this( Optional.of( id ), name, title, description, entityTypeId );
    }

    @JsonProperty( SerializationConstants.ID_FIELD )
    public UUID getId() {
        return id;
    }

    @JsonProperty( SerializationConstants.NAME )
    public String getName() {
        return name;
    }

    @JsonProperty( SerializationConstants.TITLE_FIELD )
    public String getTitle() {
        return title;
    }

    @JsonProperty( SerializationConstants.DESCRIPTION_FIELD )
    public String getDescription() {
        return description;
    }

    @JsonProperty( SerializationConstants.ENTITY_TYPE_ID )
    public UUID getEntityTypeId() {
        return entityTypeId;
    }

    @Override
    public boolean equals( Object o ) {
        if ( this == o )
            return true;
        if ( o == null || getClass() != o.getClass() )
            return false;
        CollectionTemplateType that = (CollectionTemplateType) o;
        return Objects.equals( id, that.id ) &&
                Objects.equals( name, that.name ) &&
                Objects.equals( title, that.title ) &&
                Objects.equals( description, that.description ) &&
                Objects.equals( entityTypeId, that.entityTypeId );
    }

    @Override
    public int hashCode() {
        return Objects.hash( id, name, title, description, entityTypeId );
    }

    @Override
    public String toString() {
        return "CollectionTemplateType{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", title='" + title + '\'' +
                ", description='" + description + '\'' +
                ", entityTypeId=" + entityTypeId +
                '}';
    }
}

File: api/src/main/kotlin/com/openlattice/collections/CollectionTemplateType.kt
package com.openlattice.collections


import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonProperty
import com.openlattice.authorization.securable.AbstractSecurableObject
import com.openlattice.authorization.securable.SecurableObjectType
import com.openlattice.client.serialization.SerializationConstants

import java.util.Optional
import java.util.UUID

class CollectionTemplateType

    /**
     * Creates an collection template type with provided parameters and will automatically generate a UUID if not provided.
     *
     * A collection template type wraps an entity type with additional metadata to describe a
     * particular function within an app or common workflow.
     *
     *
     * @param id An optional UUID for the collection template type.
     * @param name The unique name of the collection template type.
     * @param title The friendly name for the collection template type.
     * @param description A description of the collection template type.
     * @param entityTypeId The id of the entity type that this collection template type wraps.
     */
    @JsonCreator
    constructor(
        @JsonProperty(SerializationConstants.ID_FIELD) id: Optional<UUID>,
        @JsonProperty(SerializationConstants.NAME_FIELD) var name: String,
        @JsonProperty(SerializationConstants.TITLE_FIELD) title: String,
        @JsonProperty(SerializationConstants.DESCRIPTION_FIELD) description: Optional<String>,
        @JsonProperty(SerializationConstants.ENTITY_TYPE_ID) val entityTypeId: UUID
    ): AbstractSecurableObject(id, title, description) {

    constructor(
            id: UUID,
            name: String,
            title: String,
            description: Optional<String>,
            entityTypeId: UUID) : this(Optional.of<UUID>(id), name, title, description, entityTypeId)

    @JsonIgnore
    override fun getCategory(): SecurableObjectType {
        return SecurableObjectType.CollectionTemplateType
    }
}
